### PR TITLE
Report exit code from user-data instead of from tee

### LINKED ISSUE #9: Exit code from tee is reported
I think that this https://github.com/mcrute/tiny-ec2-bootstrap/blob/07a0b9f3c881f3efdc7ef31aa179047be835262a/tiny-ec2-bootstrap#L62

report exit code from tee and not from the user-data.sh. Am I wrong?

I can attemp to make a patch if there is a willingness to merge it.

### PR DESCRIPTION
**Pull request by graywolf-at-work**
Fixes #9

### CODE REVIEW

File: tiny-ec2-bootstrap [hunk 1]
```diff
@@ -53,15 +53,26 @@ _set_ssh_keys() {
...
-    if echo "$user_data" | grep -q '^#!/'; then
-        echo "$user_data" > /var/lib/cloud/user-data.sh
-        chmod +x  /var/lib/cloud/user-data.sh
-        /var/lib/cloud/user-data.sh 2>&1 | tee /var/log/cloud-bootstrap.log
-        echo "User Data Script Exit Status: $?"
+    if printf '%s' "$user_data" | head -n1 | grep -q '^#!/'; then
+        printf '%s' "$user_data" >/var/lib/cloud/user-data.sh
+        chmod +x /var/lib/cloud/user-data.sh
+
+        outerr=/tmp/user-data.outerr.$$
```
Comment (tomalok) [inline tiny-ec2-bootstrap:62]: I'd prefer `mktemp` over `$$`.

Does this temporary file get cleaned up anywhere?

File: tiny-ec2-bootstrap [hunk 2]
```diff
@@ -53,15 +53,26 @@ _set_ssh_keys() {
...
+        mkfifo "$outerr"
+        /var/lib/cloud/user-data.sh >"$outerr" 2>&1 &
+        u_pid=$!
+
+        tee <"$outerr" /var/log/cloud-bootstrap.log &
```
Comment (tomalok) [inline tiny-ec2-bootstrap:67]: Do we need to / should we capture `tee`'s PID and `wait` on that too?

**Comment by mcrute** [comment]
Thanks for the pull request! I agree with @tomalok 's comments. Additionally, this PR seems to add a lot of complexity with background processes and waiting. This code seems to be functionally identical with less complexity, unless I'm missing something. What do you think?

```
outfile=$(mktemp)
trap "rm $outfile" EXIT

{ /var/lib/cloud/user-data.sh; echo $? > $outfile; } | tee test.log
ec=$(cat $outfile)
```

File: tiny-ec2-bootstrap [hunk 3]
```diff
@@ -53,15 +53,26 @@ _set_ssh_keys() {
     done
 }
 
-_run_userdata() {
+_run_userdata() (
```
Comment (mcrute) [inline tiny-ec2-bootstrap:56]: What's the purpose of this change?

File: tiny-ec2-bootstrap [hunk 4]
```diff
@@ -53,15 +53,26 @@ _set_ssh_keys() {
...
     user_data=$(_get_metadata user-data)
-    if echo "$user_data" | grep -q '^#!/'; then
-        echo "$user_data" > /var/lib/cloud/user-data.sh
-        chmod +x  /var/lib/cloud/user-data.sh
-        /var/lib/cloud/user-data.sh 2>&1 | tee /var/log/cloud-bootstrap.log
-        echo "User Data Script Exit Status: $?"
+    if printf '%s' "$user_data" | head -n1 | grep -q '^#!/'; then
+        printf '%s' "$user_data" >/var/lib/cloud/user-data.sh
```
Comment (mcrute) [inline tiny-ec2-bootstrap:59]: Why use `printf` instead of `echo` here?

File: tiny-ec2-bootstrap (hunk 2 above)
Comment (mcrute) [reply tiny-ec2-bootstrap:67]: Agree. Also, please `trap` to remove any temp files created.

**Comment by graywolf-at-work** [comment]
> Thanks for the pull request! I agree with @tomalok 's comments. Additionally, this PR seems to add a lot of complexity with background processes and waiting. This code seems to be functionally identical with less complexity, unless I'm missing something. What do you think?
> 
> ```shell
> outfile=$(mktemp)
> trap "rm $outfile" EXIT
> 
> { /var/lib/cloud/user-data.sh; echo $? > $outfile; } | tee test.log
> ec=$(cat $outfile)
> ```

Afaiu this still report tee's exit code to the "outside", so eend will alway get 0. But tbh I'm not completely sure, will run some experiments (and consult POSIX spec) and get back to you.

File: tiny-ec2-bootstrap (hunk 4 above)
Comment (graywolf-at-work) [reply tiny-ec2-bootstrap:59]: Echo is not really specified in POSIX that well: https://www.etalabs.net/sh_tricks.html

So I think it's better to use printf with user input.

**Comment by graywolf-at-work** [comment]
Ok so to sum up:

```
$ { false; echo $?; } | tee /dev/null
1
```

this does work, but has this issue:

```
$ { false; echo $?; } | tee /dev/null ; echo $?
1
0
```

The tee's result is returned, so `eend` always gets the 0.

Naive solution does not work:

```
$ { false; x=$?; echo $x; } | tee /dev/null ; echo x$x
1
x
```

Since the part before `|` is run in subshell (so the variable cannot propagate). I basically see 2 options:

1. Store the return code in file (`/var/lib/cloud/user-data.exit`) and use that with `return $(cat /var/lib/cloud/user-data.exit)`.
2. Add there the trap for the cleanup and use current version.

What is the preference? Or any other option I've missed (very well possible)?

File: tiny-ec2-bootstrap (hunk 3 above)
Comment (graywolf-at-work) [reply tiny-ec2-bootstrap:56]: To not leak the `u_pid` and `outerr` variables. Performance impact should be minimal (1 fork). If you are ok with ignoring POSIX and targeting ash and bash specifically, `local` could be also an option I guess.

File: tiny-ec2-bootstrap (hunk 2 above)
Comment (graywolf-at-work) [reply tiny-ec2-bootstrap:67]: > Should we capture tee's PID and wait on that too?

I don't think that is necessasry, it dies when user script terminates and the pipe is closed. But sure, I can add it if you prefer it that way.

File: tiny-ec2-bootstrap (hunk 3 above)
Comment (tomalok) [reply tiny-ec2-bootstrap:56]: `tiny-ec2-bootstrap` depends on just the `ash` shell from busybox.

**Comment by tomalok** [comment]
> Since the part before `|` is run in subshell (so the variable cannot propagate). I basically see 2 options:
> 
> 1. Store the return code in file (`/var/lib/cloud/user-data.exit`) and use that with `return $(cat /var/lib/cloud/user-data.exit)`.
> 2. Add there the trap for the cleanup and use current version.

a test from the command line...
```
a$ sh
b$ outfile="$(mktemp)"
b$ trap "rm $outfile" EXIT
b$ { false; echo "$?" > "$outfile"; } | tee blah.log
b$ cat "$outfile"
1
b$ exit "$(cat $outfile)"
a$ echo $?
1
```
...it would seem to me that this works and is relatively simple and straightforward.

There may be some value in saving the `user-data.sh` exit value, as we could skip `mktemp` and the `trap` altogether.

File: tiny-ec2-bootstrap (hunk 3 above)
Comment (mcrute) [reply tiny-ec2-bootstrap:56]: I'd prefer `local`. `ash` is our minimum target and it supports `local`.

File: tiny-ec2-bootstrap (hunk 4 above)
Comment (mcrute) [reply tiny-ec2-bootstrap:59]: Interesting. Thanks for the link. I'm fine with changing this to `printf`.

**Comment by mcrute** [comment]
@graywolf-at-work using `$?` between the inner statements and the outer ones will definitely not work since `tee` will replace the value of `$?`. Also there's no way that I'm aware of to export a variable out of a sub-shell so the approach of defining a variable and updating it in a sub-shell isn't really workable either.

@tomalok there are other ways to grab the exit status using file descriptor redirection but it's not terribly straightforward unless you're really used to reading fd-redirection syntax (and maybe lisp :smile:).

```
{{{{{exec 3>&- 4>&-; /var/lib/cloud/user-data.sh}; echo $? >&3} | tee test.log >&4} 3>&1} | {read xs; exit $xs}} 4>&1
```

**Comment by mcrute** [comment]
@graywolf-at-work also, my apologies, the merge conflicts are because I switched the file from spaces to tabs.

**Comment by graywolf-at-work** [comment]
> @tomalok there are other ways to grab the exit status using file descriptor redirection but it's not terribly straightforward unless you're really used to reading fd-redirection syntax (and maybe lisp smile).
> 
> ```shell
> {{{{{exec 3>&- 4>&-; /var/lib/cloud/user-data.sh}; echo $? >&3} | tee test.log >&4} 3>&1} | {read xs; exit $xs}} 4>&1
> ```

I'm torn between "cool" and "insane"... :)

**Comment by graywolf-at-work** [comment]
I've pushed rebased version. It used file to store the exit code and leaves it around for later inspection if necessary. I can add trap to remove it if it's not desired behavior.

**Comment by mcrute** [comment]
> I'm torn between "cool" and "insane"... :)

Me too :smile:

**Comment by mcrute** [comment]
@graywolf-at-work I'm happy with this solution. I think it's most correct to `trap` and cleanup the exit status file but I don't feel strongly about this. Were you able to test that this worked?

I'll give @tomalok a day to chime in if he has any thoughts otherwise I'll merge this. Thanks for the patch!

**Comment by graywolf-at-work** [comment]
> Were you able to test that this worked?

Well, good thing you asked. I did not, so I did and there was a bug. Fix was landed and push, diff for record was:

```
diff --git a/tiny-ec2-bootstrap b/tiny-ec2-bootstrap
index 9cf03be..944446d 100644
--- a/tiny-ec2-bootstrap
+++ b/tiny-ec2-bootstrap
@@ -54,7 +54,7 @@ _set_ssh_keys() {
 }

 _run_userdata() {
-       local user_data=$(_get_metadata user-data)
+       local user_data="$(_get_metadata user-data)"
        if printf '%s' "$user_data" | head -n1 | grep -q '^#!/'; then
                printf '%s' "$user_data" >/var/lib/cloud/user-data.sh
                chmod +x /var/lib/cloud/user-data.sh

```

I had no idea `local` actually changes how the assign work :-O ... Sorry about that.

**Comment by mcrute** [comment]
No problem! Thanks for testing. I'll give it a run too before release but it's nice to catch that stuff early.
